src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -0,0 +1,196 @@
+"""
+File       : MSMonitor.py
+Author     : Valentin Kuznetsov <vkuznet AT gmail dot com>
+             Alan Malta <alan dot malta AT cern dot ch >
+Description: MSMonitor class provide whole logic behind
+the transferor monitoring module.
+"""
+# futures
+from __future__ import division, print_function
+
+# system modules
+import time
+
+# WMCore modules
+from WMCore.MicroService.Unified.MSCore import MSCore
+
+
+class MSMonitor(MSCore):
+    """
+    MSMonitor class provide whole logic behind
+    the transferor monitoring module.
+    """
+    def __init__(self, msConfig, logger=None):
+        """
+        Runs the basic setup and initialization for the
+        MS Transferor module
+        :param microConfig: microservice configuration
+        """
+        super(MSMonitor, self).__init__(msConfig, logger)
+
+    def execute(self, reqStatus):
+        """
+        Executes the MS monitoring logic
+        :param reqStatus: request statue to process
+        :return:
+        """
+        try:
+            # get requests from ReqMgr2 data-service for given statue
+            # here with detail=False we get back list of records
+            requests = self.reqmgr2.getRequestByStatus([reqStatus], detail=False)
+            self.logger.debug('+++ monit found %s requests in %s state',
+                              len(requests), reqStatus)
+
+            requestStatuses = {}  # keep track of request statuses
+            for reqName in requests:
+                req = {'name': reqName, 'reqStatus': reqStatus}
+                self.logger.debug("+++ request %s", req)
+                # obtain status records from couchdb for given request
+                statusRecords = self.getStatusRecords(reqName)
+                # check status records and obtain completion status
+                statusRecords, completed = self.checkStatusRecords(statusRecords)
+                # if all transfers are completed
+                # move the request status staging -> staged
+                if completed == 100:
+                    self.logger.debug(
+                        "+++ request %s all transfers are completed", req)
+                    self.change(req, 'staged', '+++ monit')
+                # if pileup transfers are completed AND
+                # some input blocks are completed,
+                # move the request status staging -> staged
+                elif self.pileupTransfersCompleted(reqName):
```

Comment:
This check has likely to be done together with the input data.

Reply:
what IS the input data? please be explicit.

Reply:
point 2. here lists what "input data" can be:
https://github.com/dmwm/WMCore/wiki/ReqMgr2-MicroService-Transferor#relying-on-reqmgr2-second-proposal

Reply:
if I read wiki properly the point 2 tells me that input data is a list of datasets (either primary, parent, or pileup datasets). But the point is that you insisted in code to use this call
```
requests = self.reqmgr2.getRequestByStatus([reqStatus], detail=False)
```
see line https://github.com/dmwm/WMCore/pull/9276/files#diff-4cd93292a8b855fbad20e8e899adb8b7R32
This call returns ONLY request names and other information. How can I obtain the input data for a given request? Do I need to make another call? Or, should I change `getRequestByStatus` details to True to get all details? In other words if you want to use input data (list of datasets from request) then we should get them somehow.

Please let me know how you'd like to resolve this.

Reply:
That's right, "input data" is a list of datasets.
Getting requests with "detail=false" returns only the request name indeed, but that's why I said we need to then retrieve the transfer record (which is named after the workflow name), that transfer document will have all the input data related information that we need (including transfer IDs).

In short,
1. we retrieve workflow names in the `staging` status
2. then we retrieve its transfer document in order to proceed with the input dataset monitoring/check (note here there is a PS., as mentioned somewhere else in this PR about bulk get)

Reply:
ok, not it is clear. I changed input parameter of pileup method and implemented basic logic according to wiki (it may not be complete though).